lib/src/tag_detection.dart

```
@@ -0,0 +1,569 @@
+
+  _PathFinder(this.graph, this.predicate);
+
+  PathResult<T> findPath(T root) {
+    final libraryStack = <T>[];
+    final todo = <List<T>>[
+      [root]
+    ];
+    final visited = <T>{};
+
+    while (todo.isNotEmpty) {
+      final x = todo.last;
+      if (x.isEmpty) {
+        todo.removeLast();
+        if (todo.isNotEmpty) libraryStack.removeLast();
+        continue;
+      }
+      final current = x.removeLast();
+
+      if (visited.contains(current)) continue;
+      visited.add(current);
+
+      if (_cache.containsKey(current)) {
+        return _cache[root] = _cache[current].prefix(libraryStack);
+      }
+
+      libraryStack.add(current);
+      if (predicate(current)) {
+        return PathResult<T>.path(libraryStack);
+      }
+      todo.add(graph.directSuccessors(current).toList());
+    }
+    return PathResult<T>.noPath();
+  }
+}
+
+/// A graph of import/export dependencies for libraries under some configuration
+/// of declared variables.
+@visibleForTesting
+class LibraryGraph implements _DirectedGraph<Uri> {
+  final AnalysisSession _analysisSession;
+  final Map<String, String> _declaredVariables;
+  final Map<Uri, Set<Uri>> _cache = <Uri, Set<Uri>>{};
+
+  LibraryGraph(this._analysisSession, this._declaredVariables);
+
+  /// The direct successors of the library [uri] are the libraries imported or
+  /// exported by that library.
+  ///
+  /// Part files have no successors.
+  ///
+  /// For the purposes of our analysis `dart:` and `dart-ext:` libraries have no
+  /// successors.
+  @override
+  Set<Uri> directSuccessors(Uri uri) {
+    return _cache.putIfAbsent(uri, () {
+      final uriString = uri.toString();
+      if (uriString.startsWith('dart:') || uriString.startsWith('dart-ext:')) {
+        return <Uri>{};
+      }
+      final unitResult = _analysisSession.getParsedUnit(
+        _analysisSession.uriConverter.uriToPath(uri),
+      );
+      if (unitResult.isPart || unitResult.unit == null) {
```

This means we ignore errors `unitResult.unit == null`.